meilisearch-http/src/index/search.rs

```diff
@@ -291,59 +411,121 @@ impl<'a, A: AsRef<[u8]>> Highlighter<'a, A> {
                     .collect(),
             ),
             Value::Object(object) => Value::Object(
                 object
                     .into_iter()
-                    .map(|(k, v)| (k, self.highlight_value(v, words_to_highlight)))
+                    .map(|(k, v)| (k, self.format_value(v, matcher, FormatOptions { highlight: format_options.highlight, crop: None })))
                     .collect(),
             ),
+            value => value,
         }
     }
+
+    fn format_string(
+        &self,
+        s: String,
+        matcher: &impl Matcher,
+        format_options: FormatOptions,
+    ) -> String {
+        let analyzed = self.analyzer.analyze(&s);
+
+        let tokens: Box<dyn Iterator<Item = (&str, Token)>> = match format_options.crop {
+            Some(crop_len) => {
+                let mut buffer = Vec::new();
+                let mut tokens = analyzed.reconstruct().peekable();
+
+                while let Some((word, token)) = tokens.next_if(|(_, token)| matcher.matches(token.text()).is_none()) {
+                    buffer.push((word, token));
+                }
+
+                match tokens.next() {
+                    Some(token) => {
+                        let mut total_len: usize = buffer.iter().map(|(word, _)| word.len()).sum();
+                        let before_iter = buffer.into_iter().skip_while(move |(word, _)| {
+                            total_len -= word.len();
+                            total_len >= crop_len
+                        });
+
+                        let mut taken_after = 0;
+                        let after_iter = tokens
+                        .take_while(move |(word, _)| {
+                            let take = taken_after < crop_len;
+                            taken_after += word.chars().count();
+                            take
+                        });
+
+                        let iter = before_iter
+                            .chain(Some(token))
+                            .chain(after_iter);
+
+                        Box::new(iter)
+                    },
+                    // If no word matches in the attribute
+                    None => {
+                        let mut count = 0;
+                        let iter = buffer.into_iter().take_while(move |(word, _)| {
+                            let take = count < crop_len;
+                            count += word.len();
+                            take
+                        });
+
+                        Box::new(iter)
+                    }
+                }
+            }
+            None => Box::new(analyzed.reconstruct()),
+        };
+
+        tokens
+            .map(|(word, token)| {
+                if format_options.highlight && token.is_word() && matcher.matches(token.text()).is_some() {
+                    let mut new_word = String::new();
+                    new_word.push_str(&self.marks.0);
+                    if let Some(match_len) = matcher.matches(token.text()) {
```

Comment:
@curquiza Calling `matcher.matches(token.text())` two times here is not quite good, this method can be CPU intensive in the case of a short prefix in the user query for example, and remember that you call it on all of the document tokens.

Reply:
Thanks for the info @kerollmops, I'll fix that